### DISCUSSION TITLE
Custom skeleton/scaffolding and working in different environments

### DISCUSSION THREAD

**Discussion by ska-ibees** [getporter/porter#1376] [Q&A]
Hello Carolyn,

Moving slack question here...

Let's say you are working on a project and need to switch between Development, Test, Staging/Pre-production and Production Environments.

During development you may need few extra tools/utilities like pre-commits, specific linting, IDE workspace (eg. VS-Code) etc. which are not much required in other environments. 

And you want to make a template with all those tools and config files that couold be saved somewhere and could be initialzed with 
`porter create`

Based on above scenerio, could you please answer below:

1. Is there a safe way to achieve this without changing porter core code? 
2. Is there any existing possibility to have enviornment specific setup. Like by using env specific porter.yaml files? e.g. portal_dev.yaml, portal_test.yaml and portal_prod.yaml etc.

**Comment by carolynvs** [comment]
Sorry for all the questions! I'm just trying to get a better picture of your use case. 😀 

1. Can you help clarify what the user is trying to do when they run `porter create`? 
1. Are they making a new bundle and want extra files included by default _to be built into the bundle and used by the bundle during porter install_?
1. Are you making different bundles for each environment?
1. Is this bundle installing an application? Do you run it to install the application in dev, test, prod?

I ask because our recommendation is to make a bundle do different things based on a variable (such as environment) is with parameters. e.g. `porter install --param dev`. But I'm not sure what exactly the bundle is doing differently in each environment so that may be extremely cumbersome.

**Reply by ska-ibees** [reply]
> 
> 
> Sorry for all the questions! I'm just trying to get a better picture of your use case. 😀

I am astonished to see such a prompt response. Thanks again for the quailty support.

>     1. Can you help clarify what the user is trying to do when they run `porter create`?

User is starting a new project having specific needs like pre-commits, specific linting, IDE workspace (eg. VS-Code) etc.

>     2. Are they making a new bundle and want extra files included by default _to be built into the bundle and used by the bundle during porter install_?

Yes

>     3. Are you making different bundles for each environment?

Don't want to make seperate bundles for each enviornment. It would be ideal to have single bundle with possibility to have multiple setups for each enviornment. For example, during local development may not need DNS, SSL etc. 

>     4. Is this bundle installing an application? Do you run it to install the application in dev, test, prod?

Yes.

> I ask because our recommendation is to make a bundle do different things based on a variable (such as environment) is with parameters. e.g. `porter install --param dev`. But I'm not sure what exactly the bundle is doing differently in each environment so that may be extremely cumbersome.

Could you please point to further details (Doc or example code) about the use of such variable?

Thanks again!

**Reply by carolynvs** [reply]
Here is more documentation about parameters:

* https://porter.sh/author-bundles/#parameters
* https://porter.sh/parameters/

This allows you to pass in a variable, such as `porter install --param env=staging`, then access the parameter either using porter's templating, or in a bash script using the environment variable declared with the parameter.

Please give that a look and let me know if that works for you use case!

**Reply by ska-ibees** [reply]
👍  Seems promising.
